### PR TITLE
RR interval Outlier Calculation implemented

### PR DESCRIPTION
**Pull request by aungkonazim**

### CODE REVIEW

**Comment by twhnat** [comment]
Azim,
There are now conflicts on several files due to PR #54 Please resolve these to move forward.

File: cerebralcortex/data_processor/cStress.py [hunk 1]
```diff
@@ -59,20 +62,36 @@ def cStress(rdd: RDD) -> RDD:
...
     # Accelerometer Feature Computation
     accel_features = accel.map(lambda ds: (ds[0], accelerometer_features(ds[1], window_length=10.0)))
 
+    rip_corrected_and_quality = rip_corrected.join(rip_quality)
+
+
     # rip features
-    peak_valley = rip_corrected.map(lambda ds: (ds[0], rip.compute_peak_valley(rip=ds[1])))
+    peak_valley = rip_corrected_and_quality.map(lambda ds: (ds[0], rip.compute_peak_valley(rip=ds[1][0],rip_quality=ds[1][1])))
+    peak_valley.take(1)
```
Comment (nasirali1) [inline cerebralcortex/data_processor/cStress.py]: Why?

File: cerebralcortex/data_processor/signalprocessing/dataquality.py [hunk 2]
```diff
@@ -22,85 +22,345 @@
...
+    standard_rr_interval_sample = valid_rr_interval_sample[0]
+    previous_rr_int_quality = Quality.ACCEPTABLE
+
+    for i in range(1, len(valid_rr_interval_sample)-1):
+
+        rr_int_diff_last_good = abs(standard_rr_interval_sample - valid_rr_interval_sample[i])
+        rr_int_diff_prev_sample = abs(valid_rr_interval_sample[i - 1] - valid_rr_interval_sample[i])
+        rr_int_diff_next_sample = abs(valid_rr_interval_sample[i] - valid_rr_interval_sample[i + 1])
+
+        if previous_rr_int_quality == Quality.UNACCEPTABLE and rr_int_diff_last_good < criterion_beat_difference:
```
Comment (nasirali1) [inline cerebralcortex/data_processor/signalprocessing/dataquality.py]: What is "int" quality?

File: cerebralcortex/data_processor/signalprocessing/dataquality.py [hunk 3]
```diff
@@ -22,85 +22,345 @@
...
+            yield DataPoint.from_tuple(valid_rr_interval_time[i], Quality.ACCEPTABLE, valid_rr_interval_time[i])
+            previous_rr_int_quality = Quality.ACCEPTABLE
+            standard_rr_interval_sample = valid_rr_interval_sample[i]
+
+        elif previous_rr_int_quality == Quality.ACCEPTABLE and rr_int_diff_prev_sample > criterion_beat_difference:
+            yield DataPoint.from_tuple(valid_rr_interval_time[i], Quality.UNACCEPTABLE, valid_rr_interval_time[i])
+            previous_rr_int_quality = Quality.UNACCEPTABLE
+
+        else:
+            yield DataPoint.from_tuple(valid_rr_interval_time[i], Quality.UNACCEPTABLE, valid_rr_interval_time[i])
```
Comment (nasirali1) [inline cerebralcortex/data_processor/signalprocessing/dataquality.py]: What is the default yield out of for loop? For example, if there is no outlier then for loop won't be executed.

File: cerebralcortex/data_processor/signalprocessing/dataquality.py [hunk 4]
```diff
@@ -22,85 +22,345 @@
...
+
+    valid_rr_interval_sample = [i.sample for i in ecg_rr.data if i.sample >.3 and i.sample < 2]
+    valid_rr_interval_time = [i.start_time for i in ecg_rr.data if i.sample >.3 and i.sample < 2]
+    valid_rr_interval_difference = abs(np.diff(valid_rr_interval_sample))
+
+    maximum_expected_difference = 4.5*0.5*iqr(valid_rr_interval_difference)
+    maximum_artifact_difference = (np.median(valid_rr_interval_sample)-2.9*.5*iqr(valid_rr_interval_difference))/3
+    criterion_beat_difference = (maximum_expected_difference+maximum_artifact_difference)/2
+    if criterion_beat_difference < .2:
+        criterion_beat_difference = .2
```
Comment (nasirali1) [inline cerebralcortex/data_processor/signalprocessing/dataquality.py]: From line 109-113, are these constant would never change? Please comment this part of the code provide any reference? for using these values.

File: cerebralcortex/data_processor/signalprocessing/dataquality.py [hunk 5]
```diff
@@ -22,85 +22,345 @@
...
+def compute_dataquality(data: list,
+                        range_memory: list,
+                        signal_type: bool,
+                        threshold_band_loose: float,
+                        threshold_slope: float,
+                        acceptable_outlier_percent: float = .34,
+                        outlier_threshold_high: float = .9769,
+                        outlier_threshold_low: float = .004884,
+                        buffer_length: int = 3,
+                        ADC_range: int = 4095)->object:
```
Comment (nasirali1) [inline cerebralcortex/data_processor/signalprocessing/dataquality.py]: Object of? Please provide the class name as return type.

File: cerebralcortex/data_processor/signalprocessing/dataquality.py [hunk 6]
```diff
@@ -22,85 +22,345 @@
...
+    :param ecg_threshold_slope: The Slope threshold of ECG signal- No consecutive datapoints can have this
+    difference in values(expressed as percentage of ADC range)
+    :param buffer_length: This specifies the memory of the data quality computation. Meaning this number of past windows
+    will also have a role to decide the quality of the current window
+
+    :return: An Annotated Datastream of ECG Data quality specifying the time ranges when data quality was acceptable/non-acceptable
+    """
+
+    ecg_quality_stream = DataStream.from_datastream(input_streams=[datastream])
+    window_data = window(datastream.data, window_size=windowsize)
```
Comment (nasirali1) [inline cerebralcortex/data_processor/signalprocessing/dataquality.py]: please rename windowsize to window_size

**Review by twhnat — APPROVED** [review]

**Pull request merged**